transform.ml

```diff
@@ -44,65 +44,107 @@ let remove_dead_code prog entry=
   in
   remove_dead_code 0
 
-(* TODO: allow pruning of the true branch
- * we need to be able to negate expressions for that *)
+module LabelSet = Set.Make(String)
+
+let collect_labels prog =
+  let rec labels pc =
+    if pc = Array.length prog then LabelSet.empty else
+      let pc' = pc + 1 in
+      match[@warning "-4"] prog.(pc) with
+      | Label l -> LabelSet.union (LabelSet.singleton l) (labels pc')
+      | _ -> labels pc' in
+  labels 0
+
+let next_fresh_label used hint =
+  let is_fresh l = match LabelSet.find l used with
+    | exception Not_found -> true
+    | x -> false
+  in
+  if is_fresh hint then hint else
+    let l i = hint ^ "." ^ (string_of_int i) in
+    let rec next_fresh i =
+      let cur = l i in
+      if is_fresh cur then cur else next_fresh (i+1) in
+    next_fresh 0
+
+module LabelMap = Map.Make(String)
+
+(* Takes a list of globally occurring labels and a program
+ * returns a copy of the program with all labels fresh and
+ * an updated list of all occurring labels *)
+let copy_fresh global_labels prog =
+  let prog_labels = collect_labels prog in
+  let rec freshened_labels labels todo =
+    match todo with
+    | [] -> LabelMap.empty
+    | l :: tl ->
+      let fresh = next_fresh_label labels l in
+      let labels = LabelSet.add fresh labels in
```

Comment:
For clarity, could you call this `labels'` instead of shadowing the old `labels`?